aten/src/ATen/native/cuda/Normalization.cu

```diff
@@ -480,7 +480,8 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_backward_cuda(const Tensor& grad_o
   // save_mean and save_invstd, so it needs recalculated.
   const auto acc_type = at::toAccumulateType(input.scalar_type(), /*is_cuda=*/true);
   Tensor mean;
-  if (save_mean->defined()) {
+  TORCH_INTERNAL_ASSERT(save_mean->defined(), "save_mean should always be defined\n");
+  if (save_mean->numel() != 0) {
```

Comment:
Isn't it possible that the save_mean has 0 elements even in the case where you're not training?

Reply:
So for inference, when normalizing with running_mean and running_var, we are not computing save_mean.
for native_batch_norm, we are copying running_mean to save_mean, even though we are not computing them; while for cudnn_batch_norm, save_mean is left as 0-sized tensor.

Reply:
So the answer is yes?
In particular, what if the input has a dimension of size 0 or is a scalar?

Reply:
Good catch.
I think we would need to duplicate this short-cut into `_batch_norm_impl_index(_backward)` since backends doesn't really support size-0 inputs.

Reply:
https://github.com/pytorch/pytorch/blob/000e0689816596fada756234dbc4cf0a73790876/aten/src/ATen/native/Normalization.cpp#L534-L540

Reply:
fixed in this: https://github.com/pytorch/pytorch/pull/57321/commits/7222cb3d28f9a25b280ded6c0031ff45f5b8e324